enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,423 @@
+to mark themselves as "false" meaning "excluded by default, unless explicitly included".
+
+
+2) Defining an [install config api](https://github.com/openshift/installer/blob/790048166067273d34f76bea4220fa395b1cce1b/pkg/types/installconfig.go#L70) field whereby the user can opt out of specific capabilities.
+
+```yaml
+InstallConfig.Capabilities
+  Default: [include|exclude]
+  Exclude:
+  - console
+  - samples
+  Include:
+  - registry
+```
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration,
+plus any explicitly included capabilities.  Any resource that is not included by the CLUSTER_PROFILE, is not
+included via the Include stanza, and is not Excluded via the Exclude stanza, will be filtered or not based on
+the `Default` value (include or exclude).
+
+The `Default` field (carried through to the CVO) allows an Admin to choose what behavior they want for new
+capabilities that might come as part of a future upgrade (either have the new capability installed during
+the upgrade, or exclude it)
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* csi-*
+* insights
+* monitoring
+* ???
+
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    capabilities:
+      default: [include|exclude]
+      exclude:
+      - console
+      - samples
+      include:
+      - registry
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with `capability.openshift.io/$exclusions`
+and listed in the `exclude` section, or are not listed and the `default` setting is `exclude`.  In addition
+any explicitly included capability should be added, regardless of the CLUSTER_PROFILE.
+
+If a resource has multiple `capability.openshift.io` annotations, then the resource will be filtered if any
+of the annotations on the resource are matched by a configured filter.  This allows a component to define
+a `ServiceMonitor` which is part of both the `Monitoring` capability and the `Foo` capability.  If either
+of those capabilities are excluded, the `ServiceMonitor` should not be created.  (If Monitoring is excluded
+then no ServiceMonitors should be created, if only Foo is excluded, then only Foo's ServiceMonitors should
+be excluded).
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
```

Comment:
How do we want the CVO to handle these?  I think we probably want a `status.capabilities` in ClusterVersion, where the CVO can explain the choices it has made in the face of an ambiguous/incorrect `spec`.  For example, if you have an entry that is both included and excluded, I'd like the CVO to complain and not install if the capability wasn't already installed.  But I'd like the CVO to complain and not remove if the capability was already installed.

Reply:
my assumption was:

1) if you provide a conflicting config to the installer, the installer rejects it
2) if you provide a conflicting config while editing the CV, the validation rejects it (can't save the change)

So there isn't really a need to report the error at runtime/in a status field.

if you provide your own CV/edit the CV as part of installing a cluster from user provided manifests and it is invalid, well that could be rejected too (your install will fail)

Reply:
> if you provide a conflicting config while editing the CV, the validation rejects it (can't save the change)

We don't currently have an admission webhook in front of ClusterVersion.  We have spec/status distinctions so we can complain about data we don't like at the moment.  I guess we could grow an admission webhook if folks felt the spec/status approach breaks down over this change (for some reason?).  Or maybe you're recommending client-side guards in `oc` and web-console edit interfaces?

Reply:
definitely wasn't suggesting client side.  Spec/Status isn't terrible, but validation is certainly nicer than "i made those edits....why aren't they taking effect?  let me go look at status.  oh, i see"

If we did want to do it in status, i'd suggest one thing we do is go to a degraded condition when the config is invalid (with a message explaining why).

i'm also not sure *what* you actually do when the spec is in conflict.  If i both "exclude" and "include" a component, what ends up in status?  Is the component excluded, or included?  That is part of why i preferred a validation error, so we don't have to make that decision.

Reply:
updated the EP to propose a validating webhook.